span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/RateLimitingSpanFilter.java

```diff
@@ -0,0 +1,67 @@
+package org.hypertrace.core.spannormalizer.jaeger;
+
+import com.typesafe.config.Config;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import org.hypertrace.core.datamodel.Event;
+
+public class RateLimitingSpanFilter {
+
+  private static final String RATE_LIMIT_CONFIG_PATH = "rate.limit.config";
+  private static final String TENANT_ID_KEY = "tenantId";
+  private static final String GROUPING_KEY_KEY = "groupingKey";
+  private static final String MAX_SPANS_PER_MINUTE_KEY = "maxSpansPerMinute";
+  private static final long SPAN_COUNT_WINDOW = 3600; // fixed at 1minute
+
+  Map<String, Map<Long, Long>> tenantSpanCountMap = new HashMap<>();
+  Map<String, Long> tenantMaxSpansPerMinuteMap = new HashMap<>();
+  Set<String> seenAttributes = new HashSet<>();
+
+  public RateLimitingSpanFilter(Config config) {
+    for (Config rateLimitConfig : config.getConfigList(RATE_LIMIT_CONFIG_PATH)) {
+      String tenantId = rateLimitConfig.getString(TENANT_ID_KEY);
+      String groupingKey = rateLimitConfig.getString(GROUPING_KEY_KEY);
+      String key = generateKey(tenantId, groupingKey);
+      tenantMaxSpansPerMinuteMap.put(key, rateLimitConfig.getLong(MAX_SPANS_PER_MINUTE_KEY));
+      tenantSpanCountMap.put(key, Map.of(0L, 0L));
+      seenAttributes.add(groupingKey);
+    }
+  }
+
+  public boolean shouldDropSpan(String tenantId, Event event) {
+    if (seenAttributes.stream()
+        .noneMatch(
+            attribute ->
+                tenantMaxSpansPerMinuteMap.containsKey(generateKey(tenantId, attribute)))) {
+      return false;
+    }
+
+    boolean shouldDropSpan = false;
+    for (String attribute : seenAttributes) {
```

Comment:
We are iterating over the seen attributes again. Can we instead get rid of the stream above?

Reply:
Right, updated.